File: IPWorks DTLS Samples/DTLS Echo Server/c++/dtlsserver.cpp
/*
 * IPWorks DTLS 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks DTLS in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworksdtls
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>

#include "../../include/ipworksdtls.h"
#define LINE_LEN 80

class MyDTLSServer : public DTLSServer
{
  int FireConnected(DTLSServerConnectedEventParams *e)
  {
    if (e->StatusCode == 0)
    {
      printf("\r\nClient successfully connected from: %s: %d\n", e->SourceAddr, e->SourcePort);
      printf("ConnectionId: %d\n", e->ConnectionId);
    }
    else
    {
      printf("\r\nError connecting: %d: %s\n", e->StatusCode, e->Description);
    }
    return 0;
  }

  int FireDisconnected(DTLSServerDisconnectedEventParams *e)
  {
    printf("Client disconnecting. ConnectionId: %d\n", e->ConnectionId);
    if (e->StatusCode != 0)
    {
      printf("Error while disconnecting, removing connection. Error: %d: %s\n", e->StatusCode, e->Description);
    }
    return 0;
  }

  int FireDataIn(DTLSServerDataInEventParams *e)
  {
    printf("Received packet from client. ConnectionId: %d\n", e->ConnectionId);
    printf("Echoing packet: %s\n", e->Datagram);
    this->SendText(e->ConnectionId, e->Datagram);
    return 0;
  }

  int FireSSLServerAuthentication(DTLSClientSSLServerAuthenticationEventParams *e)
  {
    if (!e->Accept)
    {
      printf("\r\nServer provided the following certificate: \n");
      printf("Subject: %s \n", e->CertSubject);
      printf("Issuer: %s \n", e->CertIssuer);
      printf("The following problems have been determined for this certificate:  %s \n", e->Status);
      printf("Would you like to continue or cancel the connection?  [y/n] \n");

      char response[LINE_LEN];

      fgets(response, LINE_LEN, stdin);
      response[strlen(response) - 1] = '\0';
      if (strcmp(response, "y") == 0)
        e->Accept = true;
    }
    return 0;
  }

  int FireSSLStatus(DTLSServerSSLStatusEventParams *e) 
  {
    printf("[SSL Status] %s\r\n", e->Message);
    return 0;
  }
};

int main(int argc, char* argv[])
{
  MyDTLSServer dtlsserver;

  char buffer[LINE_LEN];

  if (argc < 2)
  {
    fprintf(stderr, "usage: dtlsserver port cert_path [cert_pass]\n");
    fprintf(stderr, "\n");
    fprintf(stderr, "  port        the UDP port to listen on (required)\n");
    fprintf(stderr, "  cert_path   the path to the certificate file (required)\n");
    fprintf(stderr, "  cert_pass   the password to the specified certificate\n");
    fprintf(stderr, "\nExample: dtlsserver 1234 ./test.pfx test\n");
    printf("Press enter to continue.");
    getchar();
  }
  else
  {
    printf("*****************************************************************\n");
    printf("* This demo shows how to set up an echo server using DTLSServer. *\n");
    printf("*****************************************************************\n");

    dtlsserver.SetLocalPort(atoi(argv[1]));
    dtlsserver.SetSSLCertStoreType(CST_PFXFILE);
    dtlsserver.SetSSLCertStore(argv[2], strlen(argv[2]));
    if (argc > 3)
    {
      dtlsserver.SetSSLCertStorePassword(argv[3]);
    }
    
    //The default value of "*" picks the first private key in the certificate. For simplicity this demo will use that value.
    dtlsserver.SetSSLCertSubject("*");

    dtlsserver.SetDefaultIdleTimeout(120); // optional, disconnects inactive clients after 120 seconds.
    int ret_code = dtlsserver.StartListening();

    if (ret_code)
    {
      printf("Error: %i - %s\n", ret_code, dtlsserver.GetLastError());
      goto done;
    }

    printf("Listening...\n");

    while (true)
    {
      dtlsserver.DoEvents();
    }

  done:
    if (dtlsserver.GetListening())
    {
      dtlsserver.StopListening();
    }

    printf("Exiting... (press enter)\n");
    getchar();

    return 0;
  }
}

File: IPWorks DTLS Samples/DTLS Echo Server/python/dtlsserver.py
# 
# IPWorks DTLS 2024 Python Edition - Sample Project
# 
# This sample project demonstrates the usage of IPWorks DTLS in a 
# simple, straightforward way. It is not intended to be a complete 
# application. Error handling and other checks are simplified for clarity.
# 
# www.nsoftware.com/ipworksdtls
# 
# This code is subject to the terms and conditions specified in the 
# corresponding product license agreement which outlines the authorized 
# usage and restrictions.
# 

import sys
import string
from ipworksdtls import *

input = sys.hexversion < 0x03000000 and raw_input or input


def ensureArg(args, prompt, index):
    if len(args) <= index:
        while len(args) <= index:
            args.append(None)
        args[index] = input(prompt)
    elif args[index] is None:
        args[index] = input(prompt)


def fireConnected(e):
  if e.status_code == 0:
    print("\nClient successfully connected from: " + e.source_addr + ":" + str(e.source_port))
    print("ConnectionId: " + str(e.connection_id))
  else:
    print("\nError connecting: " + str(e.status_code) + ": " + e.description)

def fireDisconnected(e):
  print("Client disconnecting. ConnectionId: " + str(e.connection_id))
  if e.status_code != 0:
    print("Error while disconnecting, removing connection. Error: " + str(e.status_code) + ":" + e.description)

def fireDataIn(e):
  print("Received packet from " + str(e.connection_id))
  print("Echoing packet back to client: " + bytes.decode(e.datagram))
  dtlsserver.send_bytes(e.connection_id, e.datagram)

def fireSSLStatus(e):
  print("[SSL Status] " + e.message)

print("******************************************************************\n")
print("* This demo shows how to set up an echo server using DTLSServer. *\n")
print("******************************************************************\n")

if len(sys.argv) < 3:
  print("usage: dtlsserver.py port cert_path cert_pass")
  print("")
  print("  port:        the UDP port to listen on (required)")
  print("  cert_path:   the path to the certificate file (required)")
  print("  cert_pass:   the password to the specified certificate")
  print("\nExample: dtlsserver.py 1234 ./test.pfx test")
else:
  try:
    dtlsserver = DTLSServer()
    dtlsserver.on_connected = fireConnected
    dtlsserver.on_disconnected = fireDisconnected
    dtlsserver.on_data_in = fireDataIn
    dtlsserver.on_ssl_status = fireSSLStatus

    dtlsserver.set_local_port(int(sys.argv[1]))
    dtlsserver.set_ssl_cert_store_type(2)
    dtlsserver.set_ssl_cert_store(sys.argv[2])
    dtlsserver.set_ssl_cert_store_password(sys.argv[3])
    dtlsserver.set_ssl_cert_subject("*")
    dtlsserver.set_default_idle_timeout(120)

    dtlsserver.start_listening()
    print("Listening...")
    print("Press Ctrl-C to shutdown")


    while True:
      dtlsserver.do_events()

  except IPWorksDTLSError as e:
    print("ERROR: %s" % e.message)

  except KeyboardInterrupt:
    print("Shutdown requested...exiting")
    dtlsserver.shutdown()






